### PR TITLE
[master < T323-MAGE] Fix C++ API memory leak on Relationships()

### PR DESCRIPTION
**Pull request by antepusic**
NB: The first 19 commits are the same as the commits in #688 (already merged).

[master < Epic] PR
- [x] Check, and update documentation if necessary
- [x] Update [changelog](https://docs.memgraph.com/memgraph/changelog)
- [x] Write E2E tests
- [x] Compare the [benchmarking results](https://bench-graph.memgraph.com/) between the master branch and the Epic branch
- [x] Provide the full content or a guide for the final git message

### CODE REVIEW

File: include/mgp.hpp [hunk 1]
```diff
@@ -1814,10 +1815,12 @@ inline GraphRelationships::Iterator &GraphRelationships::Iterator::operator++()
...
 
-    if (next == nullptr) {
-      mgp::edges_iterator_destroy(out_relationships_iterator_);
-      out_relationships_iterator_ = nullptr;
+    if (next != nullptr) {
+      return *this;
     }
+
+    mgp::edges_iterator_destroy(out_relationships_iterator_);
+    out_relationships_iterator_ = nullptr;
```
Comment (antaljanosbenjamin) [inline include/mgp.hpp:1823]: You don't return anything here.

File: include/mgp.hpp [hunk 2]
```diff
@@ -650,7 +651,7 @@ class Relationship {
   std::string_view Type() const;
 
   /// @brief Returns an iterable & indexable structure of the relationship’s properties.
-  class Properties Properties() const;
+  mgp::Properties Properties() const;
```
Comment (antoniofilipovic) [inline include/mgp.hpp:654]: From now on, we don't use the Properties class, so merge the master first into this PR.

**Review by antoniofilipovic — CHANGES_REQUESTED** [review]
Adress Benjo's comment about not returning and check my comment about Properties.

File: include/mgp.hpp (hunk 1 above)
Comment (antepusic) [reply include/mgp.hpp:1823]: That’s the intended behavior!
If the iterator hasn’t found any more relationships on the current node, it doesn’t return anything. Instead, it moves on to go over the remaining nodes and return the first valid [outbound] relationship.

Comment (antepusic) [reply include/mgp.hpp:1823]: That’s the intended behavior!
If the iterator hasn’t found any more relationships on the current node, it doesn’t return anything. Instead, it moves on to go over the remaining nodes and return the first valid [outbound] relationship.

File: include/mgp.hpp (hunk 2 above)
Comment (antepusic) [reply include/mgp.hpp]: Done

File: include/mgp.hpp (hunk 1 above)
Comment (antaljanosbenjamin) [reply include/mgp.hpp:1800]: But that is a wrong behavior. Every non-void function **must** return a value. How would you check if something is returned or not? There is no way to achieve that. Therefore this function **must always** return a reference.

Comment (antaljanosbenjamin) [reply include/mgp.hpp:1800]: Ahh sorry, I missed the scope of the function and thought you are never returrning from that function. My bad!

**Review by antaljanosbenjamin — APPROVED** [review]

**Review by jbajic — APPROVED** [review]

**Review by antoniofilipovic — APPROVED** [review]
Good to go.

**Pull request merged**